mailbox/api/src/main/java/org/apache/james/mailbox/MailboxListener.java

```diff
@@ -314,65 +324,96 @@ public MetaDataHoldingEvent(MailboxSession.SessionId sessionId, User user, Mailb
 
     }
 
-    abstract class Expunged extends MetaDataHoldingEvent {
+    class Expunged extends MetaDataHoldingEvent {
 
         /**
          *
          */
         private static final long serialVersionUID = 1L;
 
-        public Expunged(MailboxSession.SessionId sessionId, User user, MailboxPath path, MailboxId mailboxId) {
+        private final Map<MessageUid, MessageMetaData> uids;
+
+        public Expunged(MailboxSession.SessionId sessionId, User user, MailboxPath path, MailboxId mailboxId, Map<MessageUid, MessageMetaData> uids) {
             super(sessionId, user, path, mailboxId);
+            this.uids = uids;
+        }
+
+        @Override
+        public List<MessageUid> getUids() {
+            return ImmutableList.copyOf(uids.keySet());
```

Comment:
it's too late, if you think the uids are mutable, you should make a defensive copy in constructor

Reply:
I fixed it by defensive copy in constructor.

Reply:
It's probably better to change the return type in MessageEvent to Set, Collection or Iterable

Reply:
:+1: for Collection